### PR TITLE
feat(Windows): show icon in titlebar and taskbar

### LINKED ISSUE #20071: Nvim.exe file on windows should be distributed with his own icon.
### Feature already in Vim?

Yes

### The feauture
Hi devs! I'm NeoVim user on Windows 11. And Windows 11 allow us to choose the console host between legacy Windows Console Host and modern Windows Terminal. So the problem is if I'm wanna choose nvim.exe to always open some files(Like files of code or .txt files) I'm see ugly .EXE icon which is improved in Windows 11 but still not look cool as NeoVim icon. 
### Temporary solution
The _temporary_ solution is use program called ResourceHacker to manually add an icon to nvim.exe and it look good. Files look like I'm choosed nvim-qt.exe to assosiate with them but when I'm double click on it opens a Windows Terminal with NeoVim.
### Some opinion
Displaying TUI apps in Windows 11 get huge improvement with throwing out legacy Windows Console Host and there is need to adaptation to that.

### PR DESCRIPTION
**Pull request by 3N4N**
Closes #20071.

### CODE REVIEW

**Comment by 3N4N** [comment]
I'm using the icon copied from `cmake.package/`. Duplication should be avoided but I don't know how.

File: src/nvim/main.c [hunk 1]
```diff
@@ -581,6 +609,23 @@ int main(int argc, char **argv)
...
+  if (hIconSmall != NULL) {
+    SendMessage(hWnd, WM_SETICON,
+                (WPARAM)ICON_SMALL, (LPARAM)hIconSmall);
+  }
+  if (hIcon != NULL) {
+    SendMessage(hWnd, WM_SETICON,
+                (WPARAM)ICON_BIG, (LPARAM)hIcon);
+  }
+  return TRUE;
+}
```
Comment (zeertzjq) [inline src/nvim/main.c:627]: Add `#ifdef`?

File: src/nvim/main.c [hunk 2]
```diff
@@ -207,6 +207,16 @@ void early_init(mparm_T *paramp)
...
 
+bool SetConsoleIcon(HWND hWnd, HICON hIconSmall, HICON hIcon);
+void load_icon_cb(char *fname, void *cookie)
+{
+  HICON *g_hVimIcon = (HICON *)cookie;
+  g_hVimIcon = LoadImage(NULL, fname,
+                         IMAGE_ICON,
+                         64, 64,
+                         LR_LOADFROMFILE | LR_LOADMAP3DCOLORS);
+}
```
Comment (zeertzjq) [inline src/nvim/main.c:218]: Add `#ifdef`?

File: src/nvim/main.c (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/main.c:627]: An this function can be static.

File: src/nvim/main.c (hunk 2 above)
Comment (zeertzjq) [reply src/nvim/main.c:218]: Wait, is this function used?

Comment (3N4N) [reply src/nvim/main.c]: Nope. Embarrassing. Removed now.

File: src/nvim/main.c (hunk 1 above)
Comment (3N4N) [reply src/nvim/main.c]: Done.

**Comment by 3N4N** [comment]
Just wanted to confirm that, implementation quality aside, it works with a clean install (meaning: the icon file is copied to the correct place and then found properly during runtime).

console neovim with icon

File: src/nvim/main.c [hunk 3]
```diff
@@ -256,6 +275,27 @@ int main(int argc, char **argv)
...
+      if (os_path_exists(vimiconpath)) {
+        g_hVimIcon = LoadImage(NULL, vimiconpath,
+                               IMAGE_ICON, 64, 64,
+                               LR_LOADFROMFILE | LR_LOADMAP3DCOLORS);
+        SetConsoleIcon(g_hWnd, g_hVimIcon, g_hVimIcon);
+      }
+    }
+    xfree(vimiconpath);
+  }
+#endif
```
Comment (lewis6991) [inline src/nvim/main.c:297]: Can this be factored into a function.

File: src/nvim/main.c [hunk 4]
```diff
@@ -207,6 +207,25 @@ void early_init(mparm_T *paramp)
...
+#ifdef MSWIN
+static bool SetConsoleIcon(HWND hWnd, HICON hIconSmall, HICON hIcon)
+{
+  if (hWnd == NULL) {
+    return false;
+  }
+
+  if (hIconSmall != NULL) {
+    SendMessage(hWnd, WM_SETICON,
+                (WPARAM)ICON_SMALL, (LPARAM)hIconSmall);
```
Comment (justinmk) [inline src/nvim/main.c:219]: nit: can be 1 line. same for next line

**Review by justinmk — APPROVED** [review]
cool!

**Comment by justinmk** [comment]
> I'm using the icon copied from cmake.package/. Duplication should be avoided but I don't know how.

can we change this line to pull from runtime/neovim.ico ? https://github.com/neovim/neovim/blob/719afa2988825dfc343692fbac68cb4f326b09e4/cmake.packaging/CMakeLists.txt#L35

File: src/nvim/main.c (hunk 3 above)
Comment (3N4N) [reply src/nvim/main.c:288]: Done.

File: src/nvim/main.c (hunk 4 above)
Comment (3N4N) [reply src/nvim/main.c]: Done.

File: runtime/CMakeLists.txt [hunk 5]
```diff
@@ -114,6 +114,10 @@ if(NOT APPLE)
   install_helper(
     FILES ${CMAKE_CURRENT_SOURCE_DIR}/nvim.png
     DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/128x128/apps)
+
+  install_helper(
+    FILES ${CMAKE_CURRENT_SOURCE_DIR}/neovim.ico
+    DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/nvim/runtime)
```
Comment (justinmk) [inline runtime/CMakeLists.txt:120]: I wonder why this and the "nvim.png" case above, are in (NOT APPLE) block. Doesn't hurt to install them on all systems, and avoids some unnecessary differences.

File: src/nvim/main.c [hunk 6]
```diff
@@ -207,6 +207,33 @@ void early_init(mparm_T *paramp)
...
+void SetWin32ConsoleIcon()
+{
+  HWND hWnd = GetConsoleWindow();
+  const char *vimruntime = os_getenv("VIMRUNTIME");
+  if (vimruntime != NULL) {
+    size_t len = strlen(vimruntime) + strlen(_PATHSEPSTR "neovim.ico") + 1;
+    char *vimiconpath = xmalloc(len);
+    if (vimiconpath != NULL) {
+      snprintf(vimiconpath, len, "%s" _PATHSEPSTR "neovim.ico", vimruntime);
+      if (os_path_exists(vimiconpath)) {
```
Comment (justinmk) [inline src/nvim/main.c:221]:
```suggestion
      if (!os_path_exists(vimiconpath)) {
        WLOG("neovim.ico not found: %s", NameBuff);
      }
```

File: src/nvim/main.c [hunk 7]
```diff
@@ -207,6 +207,33 @@ void early_init(mparm_T *paramp)
...
 }
 
+#ifdef MSWIN
+/// Set Windows console icon
+void SetWin32ConsoleIcon()
+{
+  HWND hWnd = GetConsoleWindow();
+  const char *vimruntime = os_getenv("VIMRUNTIME");
+  if (vimruntime != NULL) {
+    size_t len = strlen(vimruntime) + strlen(_PATHSEPSTR "neovim.ico") + 1;
```
[the remark this answers is not in the document]
Comment (3N4N) [reply src/nvim/main.c]: Done.

File: src/nvim/main.c (hunk 6 above)
Comment (3N4N) [reply src/nvim/main.c]: Done.

File: src/nvim/main.c [hunk 8]
```diff
@@ -207,6 +207,30 @@ void early_init(mparm_T *paramp)
   init_signs();
 }
 
+#ifdef MSWIN
+/// Set Windows console icon
+void SetWin32ConsoleIcon()
+{
+  HWND hWnd = GetConsoleWindow();
```
Comment (zeertzjq) [inline src/nvim/main.c:214]: `NULL` check?

**Review by justinmk — APPROVED** [review]

File: src/nvim/main.c (hunk 8 above)
Comment (3N4N) [reply src/nvim/main.c]: I moved some lines around. I "think" it's okay, null-checked now.

**Comment by 3N4N** [comment]
It doesn't revert to the original icon after exiting from neovim. Neovim logo persists after exit.

Fixing now.

**Comment by 3N4N** [comment]
Supporting `:h titlestring` in win32 console was a tiny change. I decided to push it here. The context seemed similar enough.

**Comment by justinmk** [comment]
Thanks @3N4N . Small followup in https://github.com/neovim/neovim/pull/20634 , can you try that PR?

**Comment by 3N4N** [comment]
Thanks for the merge. This one was fun.
And yes, I'll report on PR-20634 shortly.

**Pull request merged**